### DISCUSSION TITLE
FreeDOS kernel enhancements for Win 3x

### DISCUSSION THREAD

**Discussion by andrewbird** [dosemu2/fdpp#178] [Ideas]
I'm not sure if there's anything useful for FDPP here, but it's reportedly helping FreeDOS kernel to run Win 3x in all modes
https://sourceforge.net/p/freedos/mailman/message/37326256/

**Comment by stsp** [comment]
Not sure either.
We already have win3x running, modulo
some kvm bugs.
Do they have smaller patches in git to
maybe cherry-pick the most obvious ones?

**Reply by andrewbird** [reply]
I haven't seen anything appear in git yet. Apparently the changes were first tried out in a development kernel, so I read the commit log on the UNSTABLE branch but found no mention of anything seemingly related. Back then merges were time consuming so I doubt nice little patches per feature or fix exist, but perhaps now it's being looked again breaking it up into understandable commits is a possibility.

**Comment by stsp** [comment]
I can't parse what you say. :)
Do you mean there is nothing in their
git, but there is some external mega-patch,
and a kernel built on top of it in-house,
only for the video for now? Hmm, doesnt
sound like the usual practice.

**Reply by andrewbird** [reply]
There nothing in public git yet that I can find (except some small patches based on FDPP changes), there is the [patch](http://server2.fdos.org/tests/kernel-win3-patch.diff) mentioned in the mail message I linked to. But as I understand this is reworked from some early 2000's FreeDOS development activity which was never merged back into master. I couldn't find this original development work, but it predates my interest in FreeDOS development so my knowledge of where to look is limited. I expect that the patch above is just being tidied prior to inclusion in master, but whether there is one commit or a series I am not sure.

**Reply by stsp** [reply]
So do you imply that the kernel
build on video cannot be reproduced
from the git sources?
Hmm, indeed they seem to cherry-pick
a few patches targeted towards various
games support. I didn't know they are
serious about cherry-picking from fdpp. :)
This would make me less shy to eventually
merge all their recent changes. :)

**Reply by stsp** [reply]
Btw, I think they found an interesting
strategy for cherry-picking.
Just looking into fdpp change-log
gives you nothing. You can't cherry-pick
pretty much anything, based on fdpp
changelog, as when fdpp was not stable,
the changelog was in flux too. In fact,
they tried to cherry-pick using the changelog
method, and some of that result can
be found in their old, abandoned git.
All these cherry-picks were wrong, and
made the builds not to work.
But this time, they seem to be the
smart-asses: they look into fdpp tickets,
and cherry-pick whatever is related to
the ticket. Which gives hell more
accuracy for the cherry-picks, as the
tickets were referenced very accurately
almost since the fdpp creation.
Good strategy, congrats. :)

**Reply by andrewbird** [reply]
> So do you imply that the kernel
build on video cannot be reproduced
from the git sources?

No I don't mean to imply this, I presume that current git master + the posted patch would be enough, but I'm guessing as I didn't try it.

**Reply by stsp** [reply]
Ok, so that patch contains:
- hooks for int13 and int19 unrelated to win31 support.
This is the major part of the patch.
Not sure if we need that or not.
- For int2f switch to local stack.
Not sure why that was needed, likely
because freedos's C handlers take
too much of a stack space to work
on a caller's stack. Not a problem
with fdpp.
- A few cosmetic amends to fdpp changes.
I applied them to git now.
- Critical section fixes.
Might be needed for win31, but under fdpp
no one hooks the relevant calls. Would be
good to cherry-pick for overall correctness.
- DOSMGR hook.
This is interesting. Maybe it has something
to do with DOS prompt support under enh mode.
- int2f/ah=0x16 hooks.
No idea why that was needed, and why
did they put them into int2F_12_handler(),
which was obviously for int2f/ah=0x12.
- int2f/ah=0x46 hooks.
Only needed for win3.0 in real mode, and
are not implemented by the patch (just
stubbed).

So once it settles into their git, we may
indeed hunt for a few parts of that patch.
Most of it is  not needed for us, and I am
not sure about why most of the changes
were needed at all (but that doesn't mean
they weren't).

**Reply by andrewbird** [reply]
I find `int 2f.1607.bx=0015.cx=03` interesting as it seems to be able to return something about the size of the cds. I'm not quite sure whether it's the entry size in bytes, the number of entries( == lastdrive), or the product of the two.

**Reply by stsp** [reply]
```
+            case 0x03:    /* get internal structure sizes */
+            {
+              if (r.CX & 0x01) /* size of Current Directory Structure in bytes */
+              {
+                r.DX = 0xA2AB;   /* on succes DS:AX set to A2AB:B97Ch */
+                r.AX = 0xB97C;
+                r.CX = sizeof(struct cds);
+              }
```

Well, even though you can find
at least 2 obvious bugs in this
code snip, its quite obvious that
the size of 1 cds entry in bytes.

**Reply by stsp** [reply]
Well, with so many bugs and
typos I wonder if we should ever
take parts of that code.

**Reply by andrewbird** [reply]
After reading the FreeDOS patch and Eric's text file I had been looking at the [rbil](http://www.ctyme.com/intr/rb-4506.htm) which I find equally difficult to read due to formatting. I guess we'll have to back up the info gleaned from the test patch with experimental data from testing other DOSes, but it's still interesting to me. From a personal view I'd really like to find a function that returned the offset of the DOS disk stack within the data segment, as I think that would be pure gold for calling new int2f/11 redirector functions from within lfn.c etc.

**Reply by stsp** [reply]
But IIRC we are not calling
redirector functions from lfn.c
directly. There are tricks to
identify lfn.c caller in mfs.c,
or sometimes the extended
DOS functions in fdpp etc.

**Reply by andrewbird** [reply]
> But IIRC we are not calling
redirector functions from lfn.c
directly.

Indeed you can't **if** the redirector function itself uses functions that need a DOS stack.

**Reply by stsp** [reply]
Why would I even want to do that?
There are 2 cases:
1. Legacy DOSes support.
This is achieved via tricks when
mfs.c identifies the lfn.c caller.
This case already works fine
because all LFN calls are mapped
to the SFN ones and the DOS
call is done (upon which mfs.c
identifies the caller and gets the
LFN by other means).
2. fdpp. With that you can call
the lfn.c funcs as redirector funcs
from DOS, and they will be able
to get CDS.

**Reply by andrewbird** [reply]
Certainly (2) is the ideal situation, but knowing that legacy DOS also has to be supported means that any new functions added by us into the redirector have to be written for this duality. An example is the new redirector function int2f/11/42 added by ecm https://github.com/dosemu2/dosemu2/commit/0759b73c168a83e3de6a58589c1aca5720bc0723, it gets called by FDPP kernel with DOS stack, but it also gets called by lfn.c for legacy DOSes with user stack. I hope for some solution where the lfn.c could setup a DOS stack too prior to calling the redirector. Obviously this code is working fine with the user stack, but say for some reason we needed to check if a name matched a loaded character device driver. DOS has a function for that [int2f/1223](http://www.ctyme.com/intr/rb-4403.htm) which we should use in the redirector but it has the DOS stack requirement, so we would end up walking the driver list obtained via LoL In my ideal world lfn.c would become a small shim that only provided the stack setup and call to the redirector for legacy DOSes. Why is this attractive? Well I think that LFN functions belong in the redirector since they are specific to each file system. If we had a second redirector, perhaps SMB or NFS, then the mangling, canonicalisation and path resolution provided by lfn.c would be incorrect for those. IMHO each redirector should provide those specialised functions.

There are probably better things to wish for than how to find the DOS stack, but I thought I'd mention why it irks me. :smile:

**Reply by stsp** [reply]
> An example is the new redirector function int2f/11/42

Yes, its an example of NEW redirector functions.
They are new not because they support LFNs,
but rather because they take SFT on input.
Let's say, they represent a design exception.
The rest of LFN support should IMHO be
done by calling the counterpart SFN int21
API which will call the redirector with DOS
stack, and then redirector should do the
detection magic, like CREATE_TRUNCATE
now does:
```
        if (!file_exists && userStack[7] == BIOSSEG &&
                            userStack[ofs] == LFN_short_name) {
          /* special case: creation of LFN's using saved path
             if original DS:{DX,SI} points to BIOSSEG:LFN_short_name */
          strcpy(fpath, lfn_create_fpath);
        }
```

... and/or for case 2 adding the NEW
redirector functions, but only for fdpp.

Does this look reasonable?

**Reply by stsp** [reply]
So in short: if we currently already
have 2 redirector functions that may
potentially be called on user's stack,
let's not multiple them. We _had to_
create those two because they take
SFT, and we were lucky they are
not interested in a DOS stack.
Let's not do that again, unless there
is a serious reason to (like some
other func needing SFT, or alike).

**Comment by andrewbird** [comment]
This patch just got added to FreeDOS https://github.com/FDOS/kernel/commit/dd91e3c005c7573a775d2ea8a01af66b00f7031d. I noticed that current fdpp code does split up the transfers. On freedos-devel there's some talk about differences in virtual environments support for it; see https://sourceforge.net/p/freedos/mailman/message/37327695/ . Is this patch or variation of it useful for FDPP?

**Reply by stsp** [reply]
I think the patch is not correct: it
removes the limit completely, whereas
I am pretty sure the 64K limit should
stay.
Anyway, I created this PR for your
considerations:
https://github.com/dosemu2/dosemu2/pull/1520

**Reply by stsp** [reply]
I added the patch to also remove
the limit for LBA I/O. No matter if
their patch is correct or not, better
to avoid new breakages.

**Reply by andrewbird** [reply]
I wonder if there's a missing part on the FreeDOS side, as I'm still seeing use of int13/02
~~~
MFS: select_drive() DS:SI = 'bigfile.com'                                       
MFS: no drive selected fn=23                                                    
INT13: ax=0201 cx=0002 dx=0181                                                  
DISK 81 read [h:1,s:1,t:0](1)->0x1840 (0184:0000)                               
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 18+128 at pos 9344
DISK read @1/0/1 (1) -> 0x1840 OK.                                              
INT13: ax=020d cx=0205 dx=0081                                                  
DISK 81 read [h:0,s:4,t:2](13)->0x2e90 (02e9:0000)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to read 13 sectors at LBA 140+128 at pos 71808
DISK read @0/2/4 (13) -> 0x2e90 OK.                                             
INT13: ax=0211 cx=0201 dx=0181                                                  
DISK 81 read [h:1,s:0,t:2](17)->0x4890 (0489:0000)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to read 17 sectors at LBA 153+128 at pos 78464
DISK read @1/2/0 (17) -> 0x4890 OK.                                             
INT13: ax=0211 cx=0201 dx=0281                                                  
DISK 81 read [h:2,s:0,t:2](17)->0x6a90 (06a9:0000)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to read 17 sectors at LBA 170+128 at pos 87168
DISK read @2/2/0 (17) -> 0x6a90 OK.                                             
INT13: ax=0211 cx=0201 dx=0381                                                  
DISK 81 read [h:3,s:0,t:2](17)->0x8c90 (08c9:0000)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to read 17 sectors at LBA 187+128 at pos 95872
DISK read @3/2/0 (17) -> 0x8c90 OK.                                             
INT13: ax=0211 cx=0301 dx=0081                                                  
DISK 81 read [h:0,s:0,t:3](17)->0xae90 (0ae9:0000)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to read 17 sectors at LBA 204+128 at pos 104576
DISK read @0/3/0 (17) -> 0xae90 OK.                                             
~~~

**Reply by stsp** [reply]
I haven't yet applied any patch.

**Reply by andrewbird** [reply]
This is with FreeDOS git + your test branch.

**Reply by stsp** [reply]
It seems to be enabling LBA support
only for certain partition types.
You need to patch kernel.asm in the
following way:
```
ForceLBA                    db 0
```
Replace with 1.

**Reply by andrewbird** [reply]
ahh, thanks. I'll give that a go.

**Reply by stsp** [reply]
If all works, please send the
LBA-enabling path to fdpp. :)

**Reply by andrewbird** [reply]
Still struggling to get FreeDOS to switch to using LBA. Just trying to hack mkfat16image now?

**Reply by andrewbird** [reply]
Hacking the partition creation worked
~~~
MFS: select_drive() DS:SI = 'bigfile.com'                                       
MFS: no drive selected fn=23                                                    
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 18](1)->0x1840 (0184:0000)                                
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 18+128 at pos 9344
DISK ext read LBA 18 (1) -> 0x1840 OK.                                          
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 140](120)->0x2e90 (02e9:0000)                             
DISK: /tmp/test-imagedir/fat16.img: Trying to read 120 sectors at LBA 140+128 at pos 71808
DISK ext read LBA 140 (120) -> 0x2e90 OK.                                       
INT13: ax=4300 cx=0004 dx=4381                                                  
DISK 81 ext write [LBA 536](120)->0x2e90 (02e9:0000)                            
DISK: /tmp/test-imagedir/fat16.img: Trying to write 120 sectors at LBA 536 at pos 274560
DISK ext write LBA 536 (120) -> 0x2e90 OK.                                      
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 260](120)->0x2e90 (02e9:0000)                             
DISK: /tmp/test-imagedir/fat16.img: Trying to read 120 sectors at LBA 260+128 at pos 133248
DISK ext read LBA 260 (120) -> 0x2e90 OK.                                       
INT13: ax=4300 cx=0004 dx=4381                                                  
DISK 81 ext write [LBA 656](120)->0x2e90 (02e9:0000)                            
DISK: /tmp/test-imagedir/fat16.img: Trying to write 120 sectors at LBA 656 at pos 336000
DISK ext write LBA 656 (120) -> 0x2e90 OK.                                      
~~~

**Reply by stsp** [reply]
This may mean some bug.
Is it similarly difficult with fdpp?

**Reply by stsp** [reply]
Oh, fdpp doesn't seem to support
LBA currently. It wasn't ported.
Looking...

**Reply by stsp** [reply]
fdpp can now use LBA.
So if everything works, please
submit the enabling patch. :)

**Reply by andrewbird** [reply]
So just for completeness I built another FreeDOS kernel without the DMA patch and ran it with the same hacked up LBA partition type. The biggest LBA transfers were only a little smaller.
~~~
 MFS: select_drive() DS:SI = 'bigfile.com'                                       
MFS: no drive selected fn=23                                                    
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 18](1)->0x1840 (0184:0000)                                
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 18+128 at pos 9344
DISK ext read LBA 18 (1) -> 0x1840 OK.                                          
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 140](104)->0x2e90 (02e9:0000)                             
DISK: /tmp/test-imagedir/fat16.img: Trying to read 104 sectors at LBA 140+128 at pos 71808
DISK ext read LBA 140 (104) -> 0x2e90 OK.                                       
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 244](1)->0x1840 (00d9:0ab0)                               
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 244+128 at pos 125056
DISK ext read LBA 244 (1) -> 0x1840 OK.                                         
INT13: ax=4200 cx=0004 dx=4281                                                  
DISK 81 ext read [LBA 245](15)->0x10090 (1009:0000)                             
DISK: /tmp/test-imagedir/fat16.img: Trying to read 15 sectors at LBA 245+128 at pos 125568
DISK ext read LBA 245 (15) -> 0x10090 OK.                                       
INT13: ax=4300 cx=0004 dx=4381                                                  
DISK 81 ext write [LBA 536](104)->0x2e90 (02e9:0000)                            
DISK: /tmp/test-imagedir/fat16.img: Trying to write 104 sectors at LBA 536 at pos 274560
DISK ext write LBA 536 (104) -> 0x2e90 OK.                                      
INT13: ax=4300 cx=0000 dx=4381                                                  
DISK 81 ext write [LBA 640](1)->0x1840 (00d9:0ab0)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to write 1 sectors at LBA 640 at pos 327808
DISK ext write LBA 640 (1) -> 0x1840 OK.                                        
~~~

**Reply by andrewbird** [reply]
I just tried your newly enabled LBA support in FDPP (obviously without the FreeDOS DMA patch) but I see an odd pattern,  just a series of single sector reads followed by a series of single sector writes.
~~~
INT13: ax=4200 cx=0000 dx=0081                                                  
DISK 81 ext read [LBA 200](1)->0x2262 (00d9:14d2)                               
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 200+128 at pos 102528
DISK ext read LBA 200 (1) -> 0x2262 OK.                                         
INT13: ax=4200 cx=0000 dx=0081                                                  
DISK 81 ext read [LBA 201](1)->0x2262 (00d9:14d2)                               
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 201+128 at pos 103040
DISK ext read LBA 201 (1) -> 0x2262 OK.                                         
INT13: ax=4200 cx=0000 dx=0081                                                  
DISK 81 ext read [LBA 202](1)->0x2262 (00d9:14d2)                               
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 202+128 at pos 103552
DISK ext read LBA 202 (1) -> 0x2262 OK.                                         
INT13: ax=4200 cx=0000 dx=0081                                                  
DISK 81 ext read [LBA 203](1)->0x2262 (00d9:14d2)                               
DISK: /tmp/test-imagedir/fat16.img: Trying to read 1 sectors at LBA 203+128 at pos 104064
DISK ext read LBA 203 (1) -> 0x2262 OK.                                         
INT13: ax=4300 cx=0000 dx=0081                                                  
DISK 81 ext write [LBA 536](1)->0x2262 (00d9:14d2)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to write 1 sectors at LBA 536 at pos 274560
DISK ext write LBA 536 (1) -> 0x2262 OK.                                        
INT13: ax=4300 cx=0000 dx=0081                                                  
DISK 81 ext write [LBA 537](1)->0x2262 (00d9:14d2)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to write 1 sectors at LBA 537 at pos 275072
DISK ext write LBA 537 (1) -> 0x2262 OK.                                        
INT13: ax=4300 cx=0000 dx=0081                                                  
DISK 81 ext write [LBA 538](1)->0x2262 (00d9:14d2)                              
DISK: /tmp/test-imagedir/fat16.img: Trying to write 1 sectors at LBA 538 at pos 275584
DISK ext write LBA 538 (1) -> 0x2262 OK.                                        
~~~

**Reply by stsp** [reply]
This is likely freedos patch is not
there. But without the patch I did,
you wouldn't even boot.

**Reply by andrewbird** [reply]
I just did a quick hack to FDPP which I think should do the same thing as the FreeDOS patch
```c
ajb@polly:/clients/common/fdpp.git$ git diff
diff --git a/kernel/dsk.c b/kernel/dsk.c
index 6c8ea7a3..363d7121 100644
--- a/kernel/dsk.c
+++ b/kernel/dsk.c
@@ -998,7 +998,8 @@ STATIC int LBA_Transfer(ddt FAR * pddt, UWORD mode, VOID FAR * buffer,
   for (; totaltodo != 0;)
   {
     /* avoid overflowing 64K DMA boundary */
-    count = DMA_max_transfer(buffer, totaltodo);
+    // count = DMA_max_transfer(buffer, totaltodo);
+    count = totaltodo;
 
     if (FP_SEG(buffer) >= 0xa000 || count == 0)
     {
```

But still getting the single sector LBA transfers.

**Reply by andrewbird** [reply]
It's comcom32! Switching to use freecom gives me the 120 sector transfers

**Reply by stsp** [reply]
```
  do
    {
    byte_count = fread(transfer_buffer, 1, 32768, source_stream);
    if (byte_count > 0)
      {
      if (fwrite(transfer_buffer, 1, byte_count, dest_stream) != byte_count)
        goto copy_error_close;
      }
    }
  while (byte_count > 0);
```

Would you like to fix that? :)

**Reply by andrewbird** [reply]
okay

**Reply by andrewbird** [reply]
So do you think that djgpp's `fread()` implementation is slicing it up into 512byte chunks?

**Reply by stsp** [reply]
Or maybe DOS.
You can't read something smaller
than 1 sector normally.
